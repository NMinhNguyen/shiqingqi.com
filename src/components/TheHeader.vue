<template>
  <div class="the-header" :style="headerStyle">
    <div class="the-header__logo">
      <logo />
    </div>
    <div class="the-header__title">
      <h2>{{ $store.getters.getText("GREETING") }}</h2>
      <h1>{{ $store.getters.getText("SITE_NAME") }}</h1>
    </div>
  </div>
</template>

<script>
import Logo from "@/components/Logo.vue";

export default {
  components: {
    Logo
  },
  props: {
    animateTitle: { type: Number, default: 0 },
    animateBackground: { type: Number, default: 0 }
  },
  computed: {
    /** @type {() => {[property: string]: number}} */
    headerStyle() {
      return {
        "--the-header-animate-title": this.animateTitle,
        "--the-header-animate-background": this.animateBackground
      };
    }
  }
};
</script>

<style lang="scss">
$animate-title: var(--the-header-animate-title);
$animate-background: var(--the-header-animate-background);

.the-header {
  position: fixed;
  top: 0;
  width: 100%;
  display: flex;
  padding: 0 1rem;
  overflow: hidden;
  align-items: center;
  z-index: $z-header;
  height: calc(6rem - #{$animate-background} * 2rem);

  &__logo {
    width: 3rem;
    height: 3rem;
    padding: 0.5rem;
    opacity: $animate-title;
    transform: translateX(calc(#{$animate-title} * 3rem - 3rem));
  }

  &__title {
    display: flex;
    flex-direction: column;
    align-self: flex-start;
    transform: translate(
      calc(#{$animate-title} * 4rem - 3rem),
      calc(#{$animate-background} * -4rem)
    );
  }

  h1,
  h2 {
    line-height: calc(6rem - #{$animate-title} * 2rem);
    font-size: calc(#{$size-l} - #{$animate-title} * (#{$size-l - $size-m}));
    margin: 0;
  }

  h1 {
    opacity: $animate-title;
  }

  h2 {
    opacity: calc(1 - #{$animate-title});
  }
}
</style>
