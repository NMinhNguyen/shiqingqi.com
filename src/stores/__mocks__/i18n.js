export default {
  state: {
    locale: "en",
    translations: {
      en: {
        test: "test"
      }
    }
  },
  mutations: {
    setLocale: jest.fn(),
    setTranslations: jest.fn()
  },
  actions: {
    fetchTranslations: jest.fn()
  },
  getters: {
    getText: jest.fn(() => () => "")
  }
};
