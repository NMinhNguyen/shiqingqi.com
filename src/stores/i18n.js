import Vue from "vue";

export const state = {
  locale: "en",
  /** @type {{[locale:string]: {[key:string]: string}}} */
  translations: {}
};

export const mutations = {
  /**
   * @param {typeof state} state State object
   * @param {string} locale Locale code
   */
  setLocale(state, locale) {
    state.locale = locale;
  },

  /**
   * @param {typeof state} state State object.
   * @param {{lang: string, translations: object}} payload New translations.
   */
  setTranslations(state, { lang, translations }) {
    Vue.set(state.translations, lang, {
      ...state.translations[lang],
      ...translations
    });
  }
};

export const actions = {
  /**
   * @param {import("vuex").ActionContext<typeof state, {}>} context
   * @param {string} lang Language code.
   * @returns {Promise}
   */
  async fetchTranslations({ commit }, lang) {
    /** @type {{default: object}} */
    const module = await import(`@/constants/translations/${lang}.json`);
    commit("setTranslations", { lang, translations: module.default });
  }
};

export const getters = {
  /**
   * @param {typeof state} state
   * @returns {(key: string, params: {[key:string]: string}) => string}
   */
  getText: state => (key, params) => {
    const translations = state.translations[state.locale];

    // Return empty string if translation doesn't exist
    if (!translations || !translations[key]) return "";

    // Return translation if no params provided
    if (!params) return translations[key];

    // Replace the params
    let text = translations[key];
    for (let name in params) {
      text = text.replace(`{{${name}}}`, params[name]);
    }
    return text;
  }
};

export default {
  state,
  mutations,
  actions,
  getters
};
