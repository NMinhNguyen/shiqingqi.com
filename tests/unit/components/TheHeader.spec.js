// @ts-nocheck
import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import TheHeader from "@/components/TheHeader.vue";
import i18n from "@/stores/i18n";

jest.mock("@/stores/i18n");

let localVue = createLocalVue();
localVue.use(Vuex);

describe("TheHeader", () => {
  let store;

  beforeAll(() => {
    store = new Vuex.Store({
      modules: {
        i18n
      }
    });
  });

  it("renders the header", () => {
    expect(
      shallowMount(TheHeader, {
        localVue,
        store
      })
    ).toMatchSnapshot();
  });
});
