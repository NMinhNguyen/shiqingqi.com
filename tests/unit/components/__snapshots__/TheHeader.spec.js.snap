// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheHeader renders the header 1`] = `
<div class="the-header">
  <div class="the-header__logo">
    <logo-stub color="#ffffff"></logo-stub>
  </div>
  <div class="the-header__title">
    <h2></h2>
    <h1></h1>
  </div>
</div>
`;
